### PR TITLE
Bench tools

### LINKED ISSUE #1984: create performance profiling scripts for python
this can be done using `cProfile` , `timeit`, ... 
See https://github.com/numenta/nupic.core/issues/28

### PR DESCRIPTION
**Pull request by breznak**
simple encoders/SP/TP benchmarks used for `cProfile` module in Python.

Fixes #1984

### CODE REVIEW

**Comment by breznak** [comment]
@scottpurdy @rhyolight before I start any work on these... these are my quick-shot code snips for profiling in python, would there be interest in including this? as `examples/profiling/..` or something, same way as we now have "profiling example in c++", or just leave it out of the main tree?

**Comment by scottpurdy** [comment]
@oxtopus - thoughts?

**Comment by oxtopus** [comment]
I think it'd be useful.  Maybe also consider a [timeit](https://docs.python.org/2/library/timeit.html) use-case

**Comment by rhyolight** [comment]
@breznak Please create an issue the describes what functionality you are suggesting at a high level and link to this PR.

File: tp_large.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+## run python -m cProfile --sort cumTime tp_large.py
```
Comment (oxtopus) [inline tp_large.py]: I get an error when I run this:

```
python -m cProfile --sort cumTime tp_large.py
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/cProfile.py", line 199, in <module>
    main()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/cProfile.py", line 192, in main
    runctx(code, globs, None, options.outfile, options.sort)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/cProfile.py", line 56, in runctx
    result = prof.print_stats(sort)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/cProfile.py", line 81, in print_stats
    pstats.Stats(self).strip_dirs().sort_stats(sort).print_stats()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pstats.py", line 225, in sort_stats
    sort_tuple = sort_tuple + sort_arg_defs[word][0]
KeyError: 'cumTime'
```

**Comment by oxtopus** [comment]
Thanks for putting this together.  I think this would be a good addition to the project, however, there are a number of issues that need to be addressed before I can merge it:

- [ ] Create a GitHub issue with a writeup for the motivation for, and usage of these tools
- [ ] Move them into a subdir, say, `utils/profiling` or something along those lines, with a README.md that includes usage details, and a guide for creating additional profiling examples.
- [ ] Run pylint on the files and fix any reported issues.  For example:

  ```
pylint tp_large.py 
************* Module tp_large
W9820,   1: 0 - Non-executable script - script has __main__ block, but the executable bit is not set. Set the executable bit on the script or remove the __main__ block. (non-executable)
C0303,  23: 0 - Trailing whitespace (trailing-whitespace)
C0303,  34: 0 - Trailing whitespace (trailing-whitespace)
C0303,  48: 0 - Trailing whitespace (trailing-whitespace)
W0612,  53: 8 - Unused variable 'i' (unused-variable)
R0201,  43: 2 - Method could be a function (no-self-use)
W0611,  29: 0 - Unused import getSeed (unused-import)
W0611,  31: 0 - Unused import SparseBinaryMatrix (unused-import)
W0611,  29: 0 - Unused import getNumpyRandomGenerator (unused-import)
W0611,  31: 0 - Unused import SparseMatrix (unused-import)
  ```

- [ ] Remove commented-out code

File: tp_large.py (hunk 1 above)
Comment (oxtopus) [inline tp_large.py]: Key should be "cumtime"

File: enc_speed.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from nupic.encoders.scalar import ScalarEncoder
+from nupic.encoders.random_distributed_scalar import RandomDistributedScalarEncoder as RDSE
+
+import numpy as np
+
+total=10000
```
Comment (oxtopus) [inline enc_speed.py]: Need spaces on either side of `=` per [coding style guide](https://github.com/numenta/nupic/wiki/Python-Style-Guide#spaces)

File: enc_speed.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+data=np.random.randint(minV, maxV+1, total)
+
+encScalar = ScalarEncoder(w=21, minval=minV, maxval=maxV, resolution=1)
+encRDSE = RDSE(resolution=1)
+
+for d in data:
+  encScalar.encode(d)
+  encRDSE.encode(d)
+
+print "Scalar n=",encScalar.n," RDSE n=",encRDSE.n
```
Comment (oxtopus) [inline enc_speed.py]: Please add spaces on either side of `,`.

File: enc_speed.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+from nupic.encoders.scalar import ScalarEncoder
```
Comment (oxtopus) [inline enc_speed.py]: Need copyright notice.

File: sp_large.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
+#! /usr/bin/env python
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
```
Comment (oxtopus) [inline sp_large.py]: `2013` -> `2015`

File: tp_large.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    """Checks that feeding in the same input vector leads to polarized
+    permanence values: either zeros or ones, but no fractions"""
+
+    tpDim = 2048
+    
+    tp = TP(numberOfCols=tpDim)
+
+    tpArray = numpy.random.randint(0, 2, tpDim).astype('float32')
+
+    for i in xrange(10000):
```
Comment (oxtopus) [inline tp_large.py]: `i` is unused.  By convention, it should be `_`.  See https://github.com/numenta/nupic/wiki/Python-Style-Guide#dummy-names

**Comment by oxtopus** [comment]
@breznak I'm done w/ review.  Please address items in https://github.com/numenta/nupic/pull/1928#issuecomment-83691003 and other inline comments and I'll merge.

Thanks!

**Comment by rhyolight** [comment]
@breznak I would still like you to create an issue for this PR as stated in my first comment.

**Comment by breznak** [comment]
@rhyolight yup, #1984

**Comment by rhyolight** [comment]
@breznak ping

**Comment by breznak** [comment]
will do, but don't have so much time for it now

**Comment by breznak** [comment]
ok, I've reworked the PR, code should be ready, what is missing is a Readme, is it necessary though? I've put good comments in the code
Please rereview @oxtopus

**Comment by oxtopus** [comment]
:+1: Thanks!

**Pull request merged**